### DISCUSSION TITLE
Set the value of machine.Pin's constructor "value" keyword argument

### DISCUSSION THREAD

**Discussion by kresimirhec** [micropython/micropython#16432] [ESP32]
I am having problems invoking machine.Pin's constructor.
I am trying to set the value of "value" keyword argument but it looks like that cannot be done.
Please check the code, it will explain more than I could do with words.
I have commented errors under each base class constructor call.

```
from machine import Pin

class Relay(Pin):
    def __init__(self, pin):
        #super().__init__(pin, Pin.OUT, None, 1)
        #TypeError: extra positional arguments given
        
        #super().__init__(pin, Pin.OUT, None, value=1)
        #TypeError: function doesn't take keyword arguments

        #super().__init__(pin, Pin.OUT, None)
        #no errors

        #super().__init__(pin, Pin.OUT, value=1)
        #TypeError: function doesn't take keyword arguments

        super().__init__(pin, Pin.OUT)
        #no errors


relay = Relay(16)
```

Here is how the constructor looks like (extracted from machine.pyi):
```
...
def __init__(
        self,
        id: Any,
        /,
        mode: int = -1,
        pull: int = -1,
        *,
        value: Any = None,
        drive: int | None = None,
        alt: int | None = None,
    ):
...
```

I could probably create some custom class and try to set this value as it suits me, or devise some hardware NOT gate to reverse the pin signal, but I am wondering:

How can one set this "value" attribute in the constructor, using the Pin class from the library?

- Board: ESP32-WROOM-32
- IDE: Thonny 4.1.6
- Micropython family: ESP32
- variant:  Espressif * ESP32 / WROOM
- version: 1.24.1

Background (irrelevant):
Even though the problem described here is not of hardware nature, here is the story on how I ran into the problem.
Relays I am working with have inverted logic, so I have to set the value to high on start,
I have tried to do a "quick fix" to avoid turning the relay on:
```
relay = Relay(16)
relay.value(1) # quick fix
```
but it was not quick enough: although the relay does not turn on, the corresponding light on relay board does flash for few milliseconds, which does not guarantee that everything will work fine.

**Comment by scruss** [comment]
MicroPython v1.24.1 on Generic ESP32 allows me to do:
```py3
import machine
p=machine.Pin(4, machine.Pin.OUT, value=1)
```
and similarly for `value=0` 

Maybe [machine.Signal()](https://docs.micropython.org/en/v1.24.0/library/machine.Signal.html) could help you. It allows you to define a pin with inverted logic

**Reply by davefes** [reply]
For your relay issue, I would suggest that you verify the startup conditions on each GPIO you want to use and then use an inverter, if necessary, so that the relay is not actuated until you tell it to.

Maybe, external pull-ups or pull-downs to help the GPIO establish the wanted startup conditions.

**Reply by kresimirhec** [reply]
> MicroPython v1.24.1 on Generic ESP32 allows me to do:
> 
> ```python
> import machine
> p=machine.Pin(4, machine.Pin.OUT, value=1)
> ```
> 
> and similarly for `value=0`

For me as well, sorry I forgot to mention that.
I realize now that saying "invoking machine.Pin's constructor" might be misleading in this case

> 
> Maybe [machine.Signal()](https://docs.micropython.org/en/v1.24.0/library/machine.Signal.html) could help you. It allows you to define a pin with inverted logic

Using Signal as a base class faces exactly the same issues as with Pin as a base class .
However, using Signal in general, will sure help me with relay's "inverted" logic... thanks!

**Reply by kresimirhec** [reply]
> For your relay issue, I would suggest that you verify the startup conditions on each GPIO you want to use and then use an inverter, if necessary, so that the relay is not actuated until you tell it to.
> 
> Maybe, external pull-ups or pull-downs to help the GPIO establish the wanted startup conditions.

Right, I was even thinking about some NOT gate.... but this problem did not look like a hardware issue for me.
If I start to experience some erratic behaviors with relay pin I will use a pull-up resistor to keep my relay off (i.e. pin high)... thanks!

**Comment by peterhinch** [comment]
I wonder if the problems result from subclassing `Pin`. You might try using composition rather than inheritance.

**Reply by kresimirhec** [reply]
Yes, I ended up doing exactly that, using composition...
```
from machine import Pin

class Relay:
    def __init__(self, pin):  
        #self.pin = Pin(pin, Pin.OUT, None, value=1)
        #no errors

        self.pin = Pin(pin, Pin.OUT, value=1)
        #no errors


relay = Relay(16)
```
...and it does work.

So the problem is related to subclassing/inheritance! 

NB: Technically, this was never showstopper for me. I just wanted to check if there is something wrong on how I use subclassing in MicroPython as opposed to subclassing in Python.

**Reply by peterhinch** [reply]
Subclassing built-in types can be problematic in MP. Subclassing user classes works fine.
